src/analyses/useAfterFree.ml

```diff
@@ -0,0 +1,188 @@
+(** An analysis for the detection of use-after-free vulnerabilities. *)
+
+open GoblintCil
+open Analyses
+
+module Spec : Analyses.MCPSpec =
+struct
+  include Analyses.DefaultSpec
+
+  let name () = "useafterfree"
+  (* module D = Lattice.Unit *)
+  module D = SetDomain.Make(ValueDomain.Blobs)
+  module C = Lattice.Unit
+
+  (** TODO: What about context-sensitivity? *)
+  let context _ _ = ()
+
+
+  (* HELPER FUNCTIONS *)
+
+  let check_exp (exp:exp) ctx =
+    let state = ctx.local in
+    match ctx.ask (Queries.EvalValue exp) with
+    | a when not (Queries.VD.is_top a) ->
+      begin match a with
+        | `Blob (v, s, t) -> if D.mem (v, s, t) state then true else false
+        | _ -> false
+      end
+    | _ -> false
+
+  let check_lval (lval:lval) ctx =
+    let state = ctx.local in
+    match lval with
+    | (Mem e, _) ->
+      begin match ctx.ask (Queries.EvalValue e) with
+        | a when not (Queries.VD.is_top a) ->
+          begin match a with
+            | `Blob (v, s, t) -> if D.mem (v, s, t) state then true else false
```

Comment:
Here, you are checking that the entire blob abstract value is in your set, including the abstraction of the  value etc...  Probably this checking should happen at a higher level such as the varinfo associated with the blob.

Reply:
Actually, come to think of it, we should not track the blobs themselves, but rather their varinfos!